Add time-series, bucketed auction totals for flip/flop/flap auctions

hexonaut opened the pull request:
Continuing on with adding time series queries. Uses the `bid_event` trigger table to produce aggregate totals of the auctions.

Sample Request:

	{
	  timeFlipBidTotals (ilkIdentifier:"ETH-A", rangeStart:"2020-01-01", rangeEnd:"2020-06-01", bucketInterval:{months:1}) {
	    nodes {
	      bucketStart,
	      bucketEnd,
	      lotEnd,
	      bidAmountEnd
	    }
	  },
	  timeFlapBidTotals (rangeStart:"2020-01-01", rangeEnd:"2020-06-01", bucketInterval:{months:1}) {
	    nodes {
	      bucketStart,
	      bucketEnd,
	      lotEnd,
	      bidAmountEnd
	    }
	  },
	  timeFlopBidTotals (rangeStart:"2020-01-01", rangeEnd:"2020-06-01", bucketInterval:{months:1}) {
	    nodes {
	      bucketStart,
	      bucketEnd,
	      lotEnd,
	      bidAmountEnd
	    }
	  },
	  timeBiteTotals (ilkIdentifier:"ETH-A", rangeStart:"2020-01-01", rangeEnd:"2020-06-01", bucketInterval:{months:1}) {
	    nodes {
	      bucketStart,
	      bucketEnd,
	      ink,
	      art,
	      tab
	    }
	  }
	}

Sample Response:

	{
	  "data": {
	    "timeFlipBidTotals": {
	      "nodes": [
	        {
	          "bucketStart": "2020-01-01T00:00:00",
	          "bucketEnd": "2020-02-01T00:00:00",
	          "lotEnd": "210000000000000000000000000000000000000000000000000",
	          "bidAmountEnd": "594631296312757266648"
	        },
	        {
	          "bucketStart": "2020-02-01T00:00:00",
	          "bucketEnd": "2020-03-01T00:00:00",
	          "lotEnd": "360000000000000000000000000000000000000000000000000",
	          "bidAmountEnd": "594631296312757266648"
	        },
	        {
	          "bucketStart": "2020-03-01T00:00:00",
	          "bucketEnd": "2020-04-01T00:00:00",
	          "lotEnd": "210000000000000000000000000000000000000000000000000",
	          "bidAmountEnd": "419044134033302441615"
	        },
	        {
	          "bucketStart": "2020-04-01T00:00:00",
	          "bucketEnd": "2020-05-01T00:00:00",
	          "lotEnd": "360000000000000000000000000000000000000000000000000",
	          "bidAmountEnd": "594631296312757266648"
	        },
	        {
	          "bucketStart": "2020-05-01T00:00:00",
	          "bucketEnd": "2020-06-01T00:00:00",
	          "lotEnd": "210000000000000000000000000000000000000000000000000",
	          "bidAmountEnd": "594631296312757266648"
	        }
	      ]
	    },
	    "timeFlapBidTotals": {
	      "nodes": [
	        {
	          "bucketStart": "2020-01-01T00:00:00",
	          "bucketEnd": "2020-02-01T00:00:00",
	          "lotEnd": "0",
	          "bidAmountEnd": "0"
	        },
	        {
	          "bucketStart": "2020-02-01T00:00:00",
	          "bucketEnd": "2020-03-01T00:00:00",
	          "lotEnd": "360000000000000000000000000000000000000000000000000",
	          "bidAmountEnd": "594631296312757266648"
	        },
	        {
	          "bucketStart": "2020-03-01T00:00:00",
	          "bucketEnd": "2020-04-01T00:00:00",
	          "lotEnd": "210000000000000000000000000000000000000000000000000",
	          "bidAmountEnd": "419044134033302441615"
	        },
	        {
	          "bucketStart": "2020-04-01T00:00:00",
	          "bucketEnd": "2020-05-01T00:00:00",
	          "lotEnd": "0",
	          "bidAmountEnd": "0"
	        },
	        {
	          "bucketStart": "2020-05-01T00:00:00",
	          "bucketEnd": "2020-06-01T00:00:00",
	          "lotEnd": "0",
	          "bidAmountEnd": "0"
	        }
	      ]
	    },
	    "timeFlopBidTotals": {
	      "nodes": [
	        {
	          "bucketStart": "2020-01-01T00:00:00",
	          "bucketEnd": "2020-02-01T00:00:00",
	          "lotEnd": "0",
	          "bidAmountEnd": "0"
	        },
	        {
	          "bucketStart": "2020-02-01T00:00:00",
	          "bucketEnd": "2020-03-01T00:00:00",
	          "lotEnd": "0",
	          "bidAmountEnd": "0"
	        },
	        {
	          "bucketStart": "2020-03-01T00:00:00",
	          "bucketEnd": "2020-04-01T00:00:00",
	          "lotEnd": "20979773702993164877587",
	          "bidAmountEnd": "5300000000000000000000000000000000000000000000000000"
	        },
	        {
	          "bucketStart": "2020-04-01T00:00:00",
	          "bucketEnd": "2020-05-01T00:00:00",
	          "lotEnd": "0",
	          "bidAmountEnd": "0"
	        },
	        {
	          "bucketStart": "2020-05-01T00:00:00",
	          "bucketEnd": "2020-06-01T00:00:00",
	          "lotEnd": "0",
	          "bidAmountEnd": "0"
	        }
	      ]
	    }
	  },
	    "timeBiteTotals": {
	      "nodes": [
	        {
	          "bucketStart": "2020-01-01T00:00:00",
	          "bucketEnd": "2020-02-01T00:00:00",
	          "ink": "96591926116178999016",
	          "art": "9274804352328571280283",
	          "tab": "9335585762494162885809782802195388855319970979163"
	        },
	        {
	          "bucketStart": "2020-02-01T00:00:00",
	          "bucketEnd": "2020-03-01T00:00:00",
	          "ink": "1438476009761977735638",
	          "art": "230395036675126672244764",
	          "tab": "233646288833723609989912975351972949919505710417139"
	        },
	        {
	          "bucketStart": "2020-03-01T00:00:00",
	          "bucketEnd": "2020-04-01T00:00:00",
	          "ink": "203268621280760625099412",
	          "art": "17498709672305973340878909",
	          "tab": "17830951819846020593457887971873586425807249575380755"
	        },
	        {
	          "bucketStart": "2020-04-01T00:00:00",
	          "bucketEnd": "2020-05-01T00:00:00",
	          "ink": "676485665261697866383",
	          "art": "88771227691300059167158",
	          "tab": "90543916507155519988863964049679061038555425484128"
	        },
	        {
	          "bucketStart": "2020-05-01T00:00:00",
	          "bucketEnd": "2020-06-01T00:00:00",
	          "ink": "1868270134577192951967",
	          "art": "242461980656147366703245",
	          "tab": "247311012272903105955986824452888593003992317204910"
	        }
	      ]
	    }
	  }
	}

hexonaut committed:
add time-series, bucketed auction totals for flip/flop/flap auctions

hexonaut commented:
Note: I will add tests shortly, but wanted to get some initial feedback.

hexonaut committed:
Merge branch 'staging' into bites_time_query

hexonaut committed:
added bite totals; added start and end totals for the flip/flap/flop auctions; use the start timestamp for the bucket to line up totals with the bites

hexonaut commented:
Some updates:

 * Added bite totals.
 * Added start and end amounts for lot and bid_amount so you can see what the auctions started and ended at.
 * Using MIN(block_timestamp) instead of MAX now so that the bucketing will line up with the bites.

With regards to the last bullet, `BITE` and `KICK` occur in the same block, but the final block for the auction may be several hours later. In the interest of lining up `bites` with the `bids` for aggregation it made sense to me to use the start of the auction time as the time bucket instead of the end.

I still need to add tests.

hexonaut committed:
added time_flip_bid_totals test

hexonaut committed:
removed unused code from flip bid totals test; kick events dont record a bid amount; added flap bid totals test

hexonaut committed:
typo fix on flaps test; added flop bid totals test

hexonaut committed:
added bite totals test

hexonaut commented:
This is ready for review and merging. I've added tests for every query.

hexonaut commented:
Note: I want to add a count() column as well before merging.

hexonaut committed:
Merge branch 'staging' into bites_time_query

hexonaut committed:
added count to all auction total queries

hexonaut committed:
dont count null entries for auction totals count

paytonrules commented:
Sorry for the long delay on this, but we are trying to play catch up on these now.

A couple things you'll need to do:
- A lot has changed since you submitted this, so you'll need to rebase to staging.
- Update the timestamp in your migration, since it's out of order with newer timestamp that got in.

I'll go through and look at the code more closely today but I thought I could give you the heads up on that right away. Also the reason for the build failure doesn't make sense at first glance. CLIENT_IPCPATH is setup on the build server and you shouldn't have to worry about it. Hopefully a rebase takes care of whatever went wrong there.

paytonrules commented on db/migrations/20200615121551_create_time_auction_totals_query.sql:
```diff
@@ -0,0 +1,182 @@
+       COALESCE(SUM(bid_amount_end), 0) AS bid_amount_end
+FROM buckets
+    LEFT JOIN bid_results ON (
+        block_timestamp >= extract(epoch FROM buckets.bucket_start) AND
```
I'm a little concerned about the performance here - I don't think postgres can use any index on bucket_start once you call extract on it. It would be worth taking a look at the explain plan.

hexonaut replied:
I think you're right on this. I'll change it to precompute the epoch.

hexonaut committed:
Merge branch 'staging' into bites_time_query

hexonaut committed:
update create_time_auction_totals_query timestamp to move it to the bottom of the list

hexonaut committed:
pre-compute epochs for better indexing

hexonaut commented:
I've updated all my PRs to match staging and adjusted the timestamps to today. I've also adjusted the sql based on your indexing comment. No reason not to pre-compute that value, and my own testing showed a sizeable performance improvement (2300 ms vs 500 ms on the same query).

paytonrules commented:
We've just done a bunch of changes in prep for a release - so unfortunately (and sorry) you've got to update again. Really appreciate the update on the query. I've been trying them out today and so far haven't seen any issues.

hexonaut committed:
Merge branch 'staging' into bites_time_query

hexonaut commented:
No problem at all. Updated all 3 PRs.

hexonaut committed:
fixes for upgrade to auctions 1.2 and for future upgrades

paytonrules commented:
h/t to @rmulhol who noticed this, but your queries are currently unbounded here specifically with time and interval. Somebody could, maliciously or not, potentially grind the server to a halt by saying start from "beginning of time" to "end of time" with interval "1 millisecond". 

It would help if you narrowed down the available inputs to something safer, probably fitting the more typical use cases - which you know better than I do.

hexonaut commented:
Is there a way to do a pre-condition calculation to limit this? You can determine the number of buckets by doing (range_end - range_start) / bucket_interval and we could set some sane limit of 1000 or something.

paytonrules commented:
Honestly I hadn't thought of anything that complicated - I was assuming you'd use an enum to narrow some of the fields. However you could use the assert statement to limit the number of buckets the way you're saying couldn't you? I like that idea better than mine. 
https://www.postgresqltutorial.com/plpgsql-assert/

paytonrules commented:
I should have mentioned we'll want to check what the postgraphile front end does in the event of an assertion if you go that route. It must have some semi-graceful error handling, in the event the database connection is lost or something like it, but we'd want to be sure it gives the user a useful error message.

hexonaut committed:
added assert limit of max 100 buckets per query

hexonaut commented:
I've added a max bucket limit of 100 to all queries. The assert returns a proper error state with a clear message to the user:

```
{
  "message": "Please limit requests to at most 100 buckets.",
  "locations": [
    {
      "line": 32,
      "column": 3
    }
  ],
  "path": [
    "timeFlipBidTotals"
  ]
}
```

paytonrules commented:
This is an awesome change - but it looks like you've got a typo in one of your migrations causing the SQL to fail. 

```
2020/11/01 14:58:26 goose run: ERROR 20201007080800_create_time_auction_totals_query.sql: failed to run SQL migration: failed to execute SQL query "CREATE FUNCTION api.time_flip_bid_totals(ilk_identifier TEXT, range_start TIMESTAMP, range_end TIMESTAMP, bucket_interval INTERVAL DEFAULT '1 day'::INTERVAL)\n    RETURNS SETOF api.time_bid_total AS\n$$\nDECLARE\n    r api.time_bid_total%rowtype;\n": pq: unterminated dollar-quoted string at or near "$$

DECLARE

    r api.time_bid_total%rowtype;

"

exit status 1

make[1]: *** [migrate] Error 1

make[1]: Leaving directory `/home/travis/gopath/src/github.com/makerdao/vdb-mcd-transformers'
```

hexonaut commented:
Must of missed that. I'll run a test.

hexonaut committed:
fix for goose compat

hexonaut commented:
Looks like Goose had an issue that my psql client didn't. Should be good now. I'll fix it in the other PRs as well.

hexonaut commented:
Moved to here to merge into beta: https://github.com/makerdao/vdb-mcd-transformers/pull/360
